### PR TITLE
WT-5749 Only pass a non-NULL WT_ITEM value when allocating a WT_UPDATE with associated data

### PR DESCRIPTION
**Pull request by keithbostic**

### CODE REVIEW

File: src/btree/col_modify.c [hunk 1]
```diff
@@ -39,22 +39,29 @@ __wt_col_modify(WT_CURSOR_BTREE *cbt, uint64_t recno, const WT_ITEM *value, WT_U
...
      * - A key/value pair to create an update with and append to the update list.
+     * - A key with no value to create a reserved or tombstone update to append to the update list.
+     *
+     * A "full update list" is distinguished from "an update" by checking whether it has a "next"
+     * update.
      */
-    WT_ASSERT(session, (value == NULL && upd_arg != NULL) || (value != NULL && upd_arg == NULL));
+    WT_ASSERT(session, ((modify_type == WT_UPDATE_RESERVE || modify_type == WT_UPDATE_TOMBSTONE) &&
+                         value == NULL && upd_arg == NULL) ||
+        (modify_type != WT_UPDATE_TOMBSTONE && ((value == NULL && upd_arg != NULL) ||
```
Comment (luke-pearson) [inline src/btree/col_modify.c:53]: Could you clarify why this isn't WT_UPDATE_RESERVE too? As you check against TOMBSTONE and RESERVED in the other scenario.

File: src/btree/col_modify.c [hunk 2]
```diff
@@ -39,22 +39,29 @@ __wt_col_modify(WT_CURSOR_BTREE *cbt, uint64_t recno, const WT_ITEM *value, WT_U
...
+                                                 (value != NULL && upd_arg == NULL))));
+
+    /* If we don't yet have a modify structure, we'll need one. */
+    WT_RET(__wt_page_modify_init(session, page));
```
Comment (luke-pearson) [inline src/btree/col_modify.c:57]: I take it you moved both of the `__wt_page_modify_init`s for consistency?

**Comment by luke-pearson** [comment]
This change looks good to me, I just have a few questions.

File: src/btree/col_modify.c (hunk 1 above)
Comment (keithbostic) [reply src/btree/col_modify.c]: My  mistake, thanks! Fixed in 20db497bf.

File: src/btree/col_modify.c (hunk 2 above)
Comment (keithbostic) [reply src/btree/col_modify.c:58]: Exactly, just to make the two function preambles as similar as possible.

**Comment by keithbostic** [comment]
@luke-pearson, my apologies, I should have been a lot more verbose about a couple of the changes. Specifically:

* All of the shuffling in the `bt_cursor.c` file is because we now always pass a `value` parameter down to the row/col-modify routines, rather than assuming we can pass `&cbt.iface.value` and the modify routines will ignore it if `modify_type` is reserved or a tombstone. There used to be 4 routines (2 row-store, 2 column-store), that called into the modify routines. Of those two, one took a value, and one assumed `&cbt.iface`. I got rid of the one that assumed the value, because all the routines now have to pass in NULL as the value if we're allocating a tombstone/reserved update. That change also included a bunch of semantic sugar to get the helper functions for truncate to have the right type signature.

* There was an odd little change at the front of col-modify to simplify a test, we were repeated some tests unnecessarily.

* Other than that, it's more obvious changes, always calling `__wt_upd_alloc_tombstone` when we allocate a tombstone, and getting rid of some variables we never actually used.

Let me know if you have any questions -- thanks!

**Pull request merged**